hathor/p2p/peer.py

```diff
@@ -60,213 +98,56 @@ class Peer:
-    retry_attempts: int     # how many retries were made
-    last_seen: float        # last time this peer was seen
-    flags: set[str]
-    source_file: str | None
-
-    def __init__(self, auto_generate_keys: bool = True) -> None:
-        self._log = logger.new()
-        self._settings = get_global_settings()
-        self.id = None
-        self.private_key = None
-        self.public_key = None
-        self.certificate = None
-        self.entrypoints = []
-        self.retry_timestamp = 0
-        self.retry_interval = 5
-        self.retry_attempts = 0
-        self.last_seen = inf
-        self.flags = set()
-        self._certificate_options: Optional[CertificateOptions] = None
-        self.source_file = None
-
-        if auto_generate_keys:
-            self.generate_keys()
-
-    def __str__(self):
-        return (
-            f'Peer(id={self.id}, entrypoints={self.entrypoints_as_str()}, retry_timestamp={self.retry_timestamp}, '
-            f'retry_interval={self.retry_interval})'
-        )
+    id: PeerId
+    _: KW_ONLY
+    entrypoints: list[Entrypoint] = field(default_factory=list)
+    retry_timestamp: int = 0   # should only try connecting to this peer after this timestamp
+    retry_interval: int = 5     # how long to wait for next connection retry. It will double for each failure
+    retry_attempts: int = 0     # how many retries were made
+    last_seen: float = inf        # last time this peer was seen
+    flags: set[str] = field(default_factory=set)
+    _source_file: str | None = None
+    _settings: HathorSettings = field(default_factory=get_global_settings)
 
     def entrypoints_as_str(self) -> list[str]:
         """Return a list of entrypoints serialized as str"""
         return list(map(str, self.entrypoints))
 
-    def merge(self, other: 'Peer') -> None:
+    def merge(self, other: Peer) -> None:
         """ Merge two Peer objects, checking that they have the same
         id, public_key, and private_key. The entrypoints are merged without
         duplicating their entries.
         """
-        assert (self.id == other.id)
-
-        # Copy public key if `self` doesn't have it and `other` does.
-        if not self.public_key and other.public_key:
-            self.public_key = other.public_key
-            self.validate()
-
-        if self.public_key and other.public_key:
-            assert (self.get_public_key() == other.get_public_key())
-
-        # Copy private key if `self` doesn't have it and `other` does.
-        if not self.private_key and other.private_key:
-            self.private_key = other.private_key
-            self.validate()
+        assert self.id == other.id
+        self._merge(other)
 
+    def _merge(self, other: Peer) -> None:
+        """Actual merge execution, must only be made after verifications."""
+        if type(self) is Peer and isinstance(other, PublicPeer):
```

Comment:
Don't you have to check for PrivatePeer too?

Reply:
`isinstance(other: PrivatePeer, PublicPeer) == True` because `class PrivatePeer(PublicPeer)`, but I will push a change that makes this not needed anymore.

Reply:
Ok. Have you pushed this change?

Reply:
Yes, all classes are disjoint and don't inherit each other anymore.

Reply:
Thanks!